src/net/tls.cc

```diff
@@ -332,6 +332,19 @@ future<tls::x509_cert> tls::x509_cert::from_file(
     });
 }
 
+// wrapper for gnutls_datum, with raii free
+struct gnutls_datum : public gnutls_datum_t {
+    gnutls_datum() {
+        data = nullptr;
+        size = 0;
+    }
+    ~gnutls_datum() {
+        if (data != nullptr) {
+            ::gnutls_free(data);
+        }
+    }
```

Comment:
Somewhat subject to slicing.

Reply:
? What do you mean?

Reply:
It has no destructor whatsoever. This is just a simple struct to ensure we can call to gnutls functions that allocate gnutls_datum_t:s and still have some nice, safe RAII handling of freeing the data, while maintaining nice error checks etc...
The "virtual" part of datum destruction is in `gnutls_free`, which will handle whatever might be in it. Our side is responsible for calling it, nothing more.

Reply:
> This is just a simple struct to ensure we can call to gnutls functions that allocate gnutls_datum_t:s and still have some nice, safe RAII handling of freeing the data, while maintaining nice error checks etc...

yeah, i see. just wanted to understand Avi's concern. and what data slicing would imply.

Reply:
If gnutls_datum as move-assigned to gnutls_datum_t, it will destroy prematurely.

It's not a big concern since the type doesn't escape this file.

Reply:
Not meant to be moved, no. If you really need it, I can add move-constructor etc, but it seems a bit overzealous...